src/org/exist/storage/txn/TransactionManager.java

```diff
@@ -1,82 +1,131 @@
- * retrieved via {@link BrokerPool#getTransactionManager()}. TransactionManager
- * provides methods to create, commit and rollback a transaction.
+ * The Transaction Manager provides methods to begin, commit, and abort
+ * transactions.
+ *
+ * This implementation of the transaction manager is non-blocking lock-free.
+ * It makes use of several CAS variables to ensure thread-safe concurrent
+ * access. The most important of which is {@link #state} which indicates
+ * either:
+ *     1) the number of active transactions
+ *     2) that the Transaction Manager is executing system
+ *         tasks ({@link #STATE_SYSTEM}), during which time no
+ *         other transactions are active.
+ *     3) that the Transaction Manager has (or is)
+ *         been shutdown ({@link #STATE_SHUTDOWN}).
+ *
+ * NOTE: the Transaction Manager may optimistically briefly enter
+ *     the state {@link #STATE_SYSTEM} to block the initiation of
+ *     new transactions and then NOT execute system tasks if it
+ *     detects concurrent active transactions.
+ *
+ * System tasks are mutually exclusive with any other operation
+ * including shutdown. When shutdown is requested, if system tasks
+ * are executing, then the thread will spin until they are finished.
  * 
- * @author wolf
+ * There's only one TransactionManager per database instance, it can be
+ * accessed via {@link BrokerPool#getTransactionManager()}.
  *
+ * @author Adam Retter <adam@evolvedbinary.com>
+ * @author wolf
  */
+@ThreadSafe
 public class TransactionManager implements BrokerPoolService {
 
-    /**
-     * Logger for this class
-     */
     private static final Logger LOG = LogManager.getLogger(TransactionManager.class);
 
-    private long nextTxnId = 0;
-
     private final BrokerPool pool;
     private final Optional<JournalManager> journalManager;
-
     private final SystemTaskManager systemTaskManager;
 
-    private final Map<Long, TxnCounter> transactions = new HashMap<>();
+    /**
+     * The next transaction id
+     */
+    private final AtomicLong nextTxnId = new AtomicLong();
 
-    private final Lock lock = new ReentrantLock();
+    /**
+     * Currently active transactions and their operations journal write count.
+     *  Key is the transaction id
+     *  Value is the transaction's operations journal write count.
+     */
+    private final ConcurrentHashMap<Long, TxnCounter> transactions = new ConcurrentHashMap<>();
 
     /**
-     * Initialize the transaction manager using the specified data directory.
+     * State for when the Transaction Manager has been shutdown.
+     */
+    private static final int STATE_SHUTDOWN = -2;
+
+    /**
+     * State for when the Transaction Manager has executing system tasks.
+     */
+    private static final int STATE_SYSTEM = -1;
+
+    /**
+     * State for when the Transaction Manager is idle, i.e. no active transactions.
+     */
+    private static final int STATE_IDLE = 0;
```

Comment:
could be initialised as `new AtomicLong(0)` so you could potentially perform a faster object identity comparison in certain conditions ? (in stead of value comparison) #notsure :-)

Reply:
@dizzzz I am not sure I follow your reasoning here. Do you want a call to discuss?